File: src/services/poller.ts
import { KalshiClient }  from '../clients/kalshi';
import { PolyClient }    from '../clients/polymarket';
import { SignalService } from './signal';
import { BotConfig, BotStatus, BtcMarketSnapshot } from '../types';
import { logger }        from '../utils/logger';

export class PollerService {
  private kalshi:   KalshiClient;
  private poly:     PolyClient;
  private signal:   SignalService;
  private timer:    NodeJS.Timeout | null = null;
  private lastBuyAt: number | null = null;

  public status: BotStatus = {
    pollingActive:     false,
    tradingEnabled:    false,
    totalSignals:      0,
    totalOrdersPlaced: 0,
    lastSnapshot:      null,
    lastSignal:        null,
  };

  constructor(private cfg: BotConfig) {
    this.kalshi  = new KalshiClient(cfg.kalshiApiBase);
    this.poly    = new PolyClient(
      cfg.polymarketClobBase,
      cfg.tradingEnabled ? process.env.POLYMARKET_PRIVATE_KEY : undefined
    );
    this.signal  = new SignalService(cfg);
    this.status.tradingEnabled = cfg.tradingEnabled;
  }

  start(): void {
    if (this.status.pollingActive) return;
    this.status.pollingActive = true;
    this.tick();
    logger.info(`Polling started (interval: ${this.cfg.pollIntervalMs}ms)`);
  }

  stop(): void {
    this.status.pollingActive = false;
    if (this.timer) clearTimeout(this.timer);
    logger.info('Polling stopped');
  }

  private tick(): void {
    if (!this.status.pollingActive) return;
    this.timer = setTimeout(async () => {
      try {
        await this.poll();
      } catch (err) {
        logger.error('Poll error:', err);
      } finally {
        this.tick();
      }
    }, this.cfg.pollIntervalMs);
  }

  private async poll(): Promise<void> {
    const snap = await this.fetchSnapshot();
    const sig  = this.signal.evaluate(snap);

    this.status.lastSnapshot = snap;
    this.status.lastSignal   = sig;

    if (sig.kind !== 'none') {
      this.status.totalSignals++;
      logger.info(
        `🔔 SIGNAL [${sig.kind}] Kalshi=${sig.kalshiYesCents}¢ Poly=${sig.polymarketYesCents}¢ ` +
        `spread=${sig.spreadCents}¢ — ${sig.reason}`
      );
    }

    if (sig.actionable && this.cfg.tradingEnabled) {
      const now = Date.now();
      const cooldownOk = !this.lastBuyAt || (now - this.lastBuyAt) / 1000 >= this.cfg.buyCooldownSecs;
      if (cooldownOk) {
        try {
          const orderId = await this.poly.placeBuy(this.cfg.polymarketTokenYes, this.cfg.tradeUsd);
          logger.info(`✅ Order placed: ${orderId}`);
          this.status.totalOrdersPlaced++;
          this.lastBuyAt = now;
        } catch (err) {
          logger.error('Order failed:', err);
        }
      } else {
        logger.info('⏳ Buy cooldown active — skipping order');
      }
    }
  }

  private async fetchSnapshot(): Promise<BtcMarketSnapshot> {
    const [{ quote: kalshiYes, status: kalshiStatus }, polyYes] = await Promise.all([
      this.kalshi.getBtcPrice(this.cfg.kalshiTicker),
      this.poly.getPrice(this.cfg.polymarketTokenYes, 'YES'),
    ]);

    const polyNo = this.cfg.polymarketTokenNo
      ? await this.poly.getPrice(this.cfg.polymarketTokenNo, 'NO')
      : null;

    const elapsed = Math.floor((Date.now() - this.cfg.marketStartTime.getTime()) / 1000);
    const kCents  = kalshiYes.priceCents;
    const pCents  = polyYes.priceCents;

    return {
      kalshiTicker:        this.cfg.kalshiTicker,
      polymarketTokenYes:  this.cfg.polymarketTokenYes,
      kalshiYes,
      kalshiStatus,
      polymarketYes: polyYes,
      polymarketNo:  polyNo,
      marketStart:   this.cfg.marketStartTime,
      snapshotAt:    new Date(),
      elapsedSecs:   elapsed,
      spreadCents:   kCents != null && pCents != null ? kCents - pCents : null,
    };
  }
}

File: src/poller.rs
use anyhow::Result;
use chrono::Utc;
use pk_core::{BtcMarketSnapshot, KalshiClient, PolyClient, SignalKind};
use pk_signal::{SignalConfig, SignalEngine};
use pk_signer::{ClobOrder, PolyWallet};
use rust_decimal::Decimal;
use tracing::{error, info, warn};

use crate::state::AppState;

pub async fn run(state: AppState, cfg: SignalConfig, poll_ms: u64) -> Result<()> {
    let kalshi = KalshiClient::new(
        std::env::var("KALSHI_API_BASE").ok(),
        std::env::var("KALSHI_API_TOKEN").ok(),
    );
    let poly = PolyClient::new(std::env::var("POLYMARKET_CLOB_BASE").ok());
    let engine = SignalEngine::new(cfg.clone());

    let wallet = if cfg.trading_enabled {
        match PolyWallet::from_env() {
            Ok(w) => {
                info!("Wallet loaded — EOA: {} | effective: {}", w.address, w.effective_address());
                Some(w)
            }
            Err(e) => {
                warn!("Could not load wallet: {e} — running in signal-only mode");
                None
            }
        }
    } else {
        None
    };

    let mut last_buy_at: Option<i64> = None;

    loop {
        // Check if polling is paused via /poll/stop
        {
            let s = state.read().await;
            if !s.state.polling_active {
                tokio::time::sleep(tokio::time::Duration::from_millis(poll_ms)).await;
                continue;
            }
        }

        match fetch_snapshot(&kalshi, &poly, &cfg).await {
            Ok(snap) => {
                let signal = engine.evaluate(&snap);

                // Update shared state
                {
                    let mut s = state.write().await;
                    s.state.last_snapshot = Some(snap.clone());
                    if signal.kind != SignalKind::None {
                        s.state.total_signals += 1;
                    }
                    s.state.last_signal = Some(signal.clone());
                }

                if signal.is_actionable() {
                    let now = Utc::now().timestamp();
                    let cooldown_ok = last_buy_at
                        .map(|t| now - t >= cfg.buy_cooldown_secs as i64)
                        .unwrap_or(true);

                    info!(
                        kind = ?signal.kind,
                        kalshi = ?signal.kalshi_yes_cents,
                        poly = ?signal.polymarket_yes_cents,
                        spread = ?signal.spread_cents,
                        reason = %signal.reason,
                        "🔔 SIGNAL"
                    );

                    if let Some(w) = &wallet {
                        if cooldown_ok {
                            match place_buy(&poly, w, &cfg, &snap.polymarket_token_yes).await {
                                Ok(order_id) => {
                                    info!("✅ Order placed: {order_id}");
                                    last_buy_at = Some(now);
                                    let mut s = state.write().await;
                                    s.state.total_orders_placed += 1;
                                }
                                Err(e) => error!("Order failed: {e}"),
                            }
                        } else {
                            info!("⏳ Cooldown active — skipping order");
                        }
                    }
                }
            }
            Err(e) => error!("Snapshot fetch failed: {e}"),
        }

        tokio::time::sleep(tokio::time::Duration::from_millis(poll_ms)).await;
    }
}

async fn fetch_snapshot(
    kalshi: &KalshiClient,
    poly: &PolyClient,
    cfg: &SignalConfig,
) -> anyhow::Result<BtcMarketSnapshot> {
    let ((k_quote, k_status), p_yes, p_no_opt) = tokio::try_join!(
        kalshi.get_btc_price(&cfg.kalshi_ticker),
        poly.get_yes_price(&cfg.polymarket_token_yes),
        async {
            if let Some(no_token) = &cfg.polymarket_token_no {
                Ok(Some(poly.get_no_price(no_token).await?))
            } else {
                Ok::<_, pk_core::PkError>(None)
            }
        },
    )?;

    let elapsed = (Utc::now() - cfg.market_start).num_seconds();

    Ok(BtcMarketSnapshot {
        kalshi_ticker: cfg.kalshi_ticker.clone(),
        polymarket_token_yes: cfg.polymarket_token_yes.clone(),
        kalshi_yes: Some(k_quote),
        kalshi_status: k_status,
        polymarket_yes: Some(p_yes),
        polymarket_no: p_no_opt,
        market_start: cfg.market_start,
        snapshot_at: Utc::now(),
        elapsed_secs: elapsed,
    })
}

async fn place_buy(
    poly: &PolyClient,
    wallet: &PolyWallet,
    cfg: &SignalConfig,
    token_id: &str,
) -> anyhow::Result<String> {
    let price_quote = poly.get_yes_price(token_id).await?;
    let price_frac = price_quote.price_cents / Decimal::from(100);
    let size = if price_frac.is_zero() {
        Decimal::ZERO
    } else {
        cfg.trade_usd / price_frac
    };

    let order = ClobOrder::market_buy(token_id, price_frac, size);
    let payload = pk_signer::sign_clob_order(wallet, &order).await?;
    let order_id = poly.place_buy(token_id, cfg.trade_usd, payload).await?;
    Ok(order_id)
}
